components/places/src/storage/history.rs

```diff
@@ -1144,13 +1144,78 @@ pub fn get_visit_page(
     Ok(HistoryVisitInfos { infos })
 }
 
+pub fn get_visit_page_with_bound(
+    db: &PlacesDb,
+    bound: i64,
+    offset: i64,
+    count: i64,
+    exclude_types: VisitTransitionSet,
+) -> Result<HistoryVisitInfosWithBound> {
+    let allowed_types = exclude_types.complement();
+    let infos = db.query_rows_and_then_named_cached(
+        "SELECT h.url, h.title, v.visit_date, v.visit_type, h.hidden
+         FROM moz_places h
+         JOIN moz_historyvisits v
+           ON h.id = v.place_id
+         WHERE ((1 << v.visit_type) & :allowed_types) != 0 AND
+               NOT h.hidden
+               AND v.visit_date <= :bound
+         ORDER BY v.visit_date DESC, v.id
+         LIMIT :count
+         OFFSET :offset",
+        rusqlite::named_params! {
+            ":allowed_types": allowed_types,
+            ":bound": bound,
+            ":count": count,
+            ":offset": offset,
+        },
+        HistoryVisitInfo::from_row,
+    )?;
+
+    if let Some(l) = infos.last() {
+        if l.timestamp == bound {
+            // all items' timestamp are equal to the previous bound
+            let offset = offset + infos.len() as i64;
+            Ok(HistoryVisitInfosWithBound {
+                infos,
+                bound,
+                offset,
+            })
+        } else {
+            let bound = l.timestamp;
+            let mut offset: i64 = 1;
+            let mut it = infos.iter().rev();
+            it.next();
+            for i in it {
```

Comment:
nit: this loop can probably be replaced with a combination of the [take_while](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.take_while) and [count](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.count) iterator methods, which might turn out to be a bit more readable ("reverse the list, take elements while the timestamps equals the new bound, count them").

Just a suggestion though; doing it in a loop like this is also fine :-)

Reply:
Nice, it looks much nicer using `take_while` and `count`, thanks!